File: ed.jl
using LinearAlgebra
using Test
using Pkg

##### This file uses code from the QuantumSimulation course assignments ######


"""
    transverse_ising_hamiltonian(N::Int, J::Real, h::Real)

Construct the Hamiltonian matrix of the spin-1/2 transverse Ising model
for a 1D chain with periodic boundary conditions.

H = -J ∑_<i,j> Z_i Z_j - h ∑_i X_i      

# Arguments
- `N::Int`: Number of spins in the chain.
- `J::Real`: Coefficient controlling the strength of the `S^z S^z` interaction.
- `h::Real`: Strength of the external magnetic field.

# Returns
- `H::AbstractMatrix`: A `2^N × 2^N` dense matrix representing the Hamiltonian
  in the computational basis.
"""
function transverse_ising_hamiltonian(N::Int, J::Real, h::Real)
    H = zeros(Float64, 2^N, 2^N)
    X = [0.0 1.0; 1.0 0.0] 
    Z = [1.0 0.0; 0.0 -1.0]
    id2 =[1.0 0.0; 0.0 1.0]

    for i in 1:N
        # construct Sz-Sz term
        if i == N
            v_z = [(j==N || j==1 ? Z : id2)  for j in 1:N]
        # periodic boundary condition (not elegant, but easily readable)
        else
            v_z = [(j==i || j==i+1 ? Z : id2)  for j in 1:N]
        end
        H += -J.*kron(v_z...)
        
        # construct Sx term
        v_x = [(j==i ? X : id2)  for j in 1:N]
        H += -h.*kron(v_x...)
    end
    return Hermitian(H)
end

function thermal_state(H::AbstractMatrix, beta::Real)
    eig = eigen(H)
    N, M = size(H)
    @assert N==M

    evals, V = eig.values, eig.vectors
    #could be shifted by GS energy if necessary (large beta)
    beta_exps = [exp(-beta*E) for E in evals]
    Z = sum(p for p in beta_exps)
    D = Diagonal(beta_exps./Z)
    energy = sum(evals[i]*beta_exps[i] for i in 1:N)/Z
    gibbs =  V * D * V'
    return gibbs, energy
end

function thermal_expectation_value(H::AbstractMatrix, A::AbstractMatrix, beta::Real)
    eig = eigen(H)
    N, M = size(H)
    @assert N==M

    evals, V = eig.values, eig.vectors
    #could be shifted by GS energy if necessary (large beta)
    beta_exps = [exp(-beta*E) for E in evals]
    Z = sum(p for p in beta_exps)
    
    expectation = 0

    for i in 1:N
        i_ket = V[:,i]
        i_bra = i_ket'
        matrix_element = i_bra * A * i_ket 
        expectation += beta_exps[i]/Z * matrix_element
    end
    return expectation
end

function ground_state(H::AbstractMatrix)
    eig = eigen(H)
    N, M = size(H)
    @assert N==M

    gs_energy, gs = eig.values[1], eig.vectors[:,1]
    return gs, gs_energy
end


function ground_state_expectation_value(H::AbstractMatrix, A::AbstractMatrix)
    eig = eigen(H)
    N, M = size(H)
    @assert N==M

    gs = eig.vectors[:,1]
    return gs'*A*gs
end

function get_transverse_ising_groundstate(N::Int, J::Real, h::Real)
    H = transverse_ising_hamiltonian(N, J, h)
    gs, gs_energy = ground_state(H)
    return gs, gs_energy
end

function get_transverse_ising_gibbsstate(N::Int, J::Real, h::Real, beta::Real)
    H = transverse_ising_hamiltonian(N, J, h)
    gibbs, gibbs_energy = thermal_state(H, beta)
    return gibbs, gibbs_energy
end

function get_spectrum(H::AbstractMatrix)
    return eigen(H).values
end

File: ed.py
"""Exact diagonalization helpers for the transverse Ising model."""

from __future__ import annotations

import numpy as np


def _kron_all(operators):
    result = operators[0]
    for operator in operators[1:]:
        result = np.kron(result, operator)
    return result


def transverse_ising_hamiltonian(N: int, J: float, h: float):
    """Construct the transverse Ising Hamiltonian with periodic boundary conditions."""
    if N <= 0:
        raise ValueError("N must be positive")

    H = np.zeros((2**N, 2**N), dtype=np.complex128)
    X = np.array([[0.0, 1.0], [1.0, 0.0]], dtype=np.complex128)
    Z = np.array([[1.0, 0.0], [0.0, -1.0]], dtype=np.complex128)
    I = np.array([[1.0, 0.0], [0.0, 1.0]], dtype=np.complex128)

    for i in range(N):
        # Sz-Sz term with periodic boundary conditions
        if i == N - 1:
            z_ops = [Z if (j == N - 1 or j == 0) else I for j in range(N)]
        else:
            z_ops = [Z if (j == i or j == i + 1) else I for j in range(N)]
        H += -J * _kron_all(z_ops)

        # Sx term
        x_ops = [X if j == i else I for j in range(N)]
        H += -h * _kron_all(x_ops)

    return H


def thermal_state(H: np.ndarray, beta: float):
    """Return the Gibbs state and thermal energy for a Hermitian matrix."""
    H = np.asarray(H, dtype=np.complex128)
    if H.shape[0] != H.shape[1]:
        raise ValueError("H must be square")

    evals, evecs = np.linalg.eigh(H)
    shifted_evals = evals - np.min(evals)
    beta_exps = np.exp(-beta * shifted_evals)
    Z = np.sum(beta_exps)
    weights = np.asarray(beta_exps / Z, dtype=np.complex128)
    gibbs = ((evecs.astype(np.complex128) * weights) @ evecs.conj().T).astype(np.complex128)
    energy = np.complex128(np.sum(evals * beta_exps) / Z)
    return gibbs, energy


def thermal_expectation_value(H: np.ndarray, A: np.ndarray, beta: float):
    """Return the thermal expectation value of A."""
    H = np.asarray(H, dtype=np.complex128)
    A = np.asarray(A, dtype=np.complex128)
    if H.shape[0] != H.shape[1]:
        raise ValueError("H must be square")

    evals, evecs = np.linalg.eigh(H)
    shifted_evals = evals - np.min(evals)
    beta_exps = np.exp(-beta * shifted_evals)
    Z = np.sum(beta_exps)

    expectation = np.complex128(0.0)
    for i in range(H.shape[0]):
        ket = evecs[:, i]
        expectation += np.complex128(beta_exps[i] / Z) * np.complex128(ket.conj().T @ A @ ket)
    return np.complex128(expectation)


def ground_state(H: np.ndarray):
    """Return the ground-state vector and energy."""
    H = np.asarray(H, dtype=np.complex128)
    if H.shape[0] != H.shape[1]:
        raise ValueError("H must be square")

    evals, evecs = np.linalg.eigh(H)
    return evecs[:, 0].astype(np.complex128), np.complex128(evals[0])


def ground_state_expectation_value(H: np.ndarray, A: np.ndarray):
    """Return the ground-state expectation value of A."""
    H = np.asarray(H, dtype=np.complex128)
    A = np.asarray(A, dtype=np.complex128)
    if H.shape[0] != H.shape[1]:
        raise ValueError("H must be square")

    _, evecs = np.linalg.eigh(H)
    gs = evecs[:, 0].astype(np.complex128)
    return np.complex128(gs.conj().T @ A @ gs)



def get_transition_generator_single(H0: np.ndarray, A: np.ndarray, beta: float, omega_max: float, sigma: float, quadrature_points=100):
    H0 = np.asarray(H0, dtype=np.complex128)
    A = np.asarray(A, dtype=np.complex128)
    if H0.ndim != 2 or H0.shape[0] != H0.shape[1]:
        raise ValueError("H0 must be square")
    if A.shape != H0.shape:
        raise ValueError("A must have the same shape as H0")
    if omega_max <= 0:
        raise ValueError("omega_max must be positive")
    if sigma <= 0:
        raise ValueError("sigma must be positive")
    if quadrature_points < 2:
        raise ValueError("quadrature_points must be at least 2")

    E, U = np.linalg.eigh(H0)
    M = np.conj(U).T @ A @ U
    T = np.zeros_like(M, dtype=float) 

    n = H0.shape[0]
    omegas = np.linspace(0,omega_max, quadrature_points)
    for i in range(n):
        for j in range(n):
            def f(omega):
                return np.exp(-2*sigma**2*(E[i]-E[j]-omega)**2)
            
            int_plus = np.trapezoid([f(w)*(1+np.exp(-beta*w))**(-1) for w in omegas], omegas)
            int_minus = np.trapezoid([f(-w)*(1+np.exp(beta*w))**(-1) for w in omegas], omegas)
            if i != j:
                T[i,j] = np.abs(M[i,j])**2* int_plus + np.abs(M)[j,i]**2* int_minus
    for i in range(n):
        T[i,i] = -sum(T[:,i])

    T *= np.sqrt(8*np.sqrt(np.pi)) * sigma

    return T


def get_transition_generator_average(H0: np.ndarray, operators, beta: float, omega_max: float, sigma: float, quadrature_points=100):
    operators = list(operators)
    if not operators:
        raise ValueError("operators must not be empty")

    return sum(
        get_transition_generator_single(H0, A, beta, omega_max, sigma, quadrature_points)
        for A in operators
    ) / len(operators)
